### ISSUE TITLE
Good design choice, do you have any idea on speeding up the rpn_head directly?

### ISSUE THREAD

**Issue by xmyqsh** [open-mmlab/mmdetection#60]
Good design choice for [Iteration pipeline](https://github.com/open-mmlab/mmdetection/blob/master/TECHNICAL_DETAILS.md#iteration-pipeline).
As for TwoStageDetector, especially the pyramid structure in FPN, lots of multi-image argsort in  rpn_head will be bottleneck when you attempt to enlarge the batchsize. This affects training speed and GPU utility dramatically as the batchsize getting larger.
The [SNIPER](https://github.com/mahyarnajibi/SNIPER) uses half-precision training and smaller input image size allows him has a larger batchsize even on a single GPU, he adopts the two stages training process instead of end-to-end to cope with this bottleneck.
I have tried to speed up the rpn_head by multi-processing or multi-threading in python directly. While, because of the existence of the fork overhead and the GIL, both of them seems not a choice.
Do you have any idea on speeding up the rpn_head directly?

**Comment by hellock** [comment]
With the described pipeline, the GPU usage is quite high in our experiments (near 100%). All operations are performed on GPUs and this may not be a bottleneck.

**Comment by xmyqsh** [comment]
Thanks for your reply!
Great job! All operations including the proposal and proposal target layer, as well as, the sort operations are performed on GPUs, except the nms operation. In your implementation of nms, you copy the proposal box tensor from GPU to CPU, then copy it from CPU back to GPU, then do nms.
I'm not sure if there is a bottleneck. 
The facebook's [maskrcnn-benchmark](https://github.com/facebookresearch/maskrcnn-benchmark) uses pybind11 between python and C++, and implements nms on Tensor directly. 
By the way, the sort can be replaced by topk.
Also, the [maskrcnn-benchmark](https://github.com/facebookresearch/maskrcnn-benchmark) uses FrozenBN.

I have tested on my two 1080Ti, the GPU usage is from 60% to 100% periodically. Do you have any idea for the difference GPU usage from my 1080Ti and your V100?

By the way, could you explain the advantage of your [Iteration pipeline](https://github.com/open-mmlab/mmdetection/blob/master/TECHNICAL_DETAILS.md#iteration-pipeline) in your sight? I notice the [maskrcnn-benchmark](https://github.com/facebookresearch/maskrcnn-benchmark) also use this design choice.

**Comment by hellock** [comment]
1. The nms operation is usually not a bottleneck when batch size is small, but if you use large batch sizes, it may take some time. This needs further profiling.
2. The nms implementation in maskrcnn-benchmark is better, and we consider using pytorch instead of cython to implement it.
3. topk is a better choice, a PR is welcome.
4. We freeze parameters as well as stats for caffe-style backbones, but for pytorch-style backbones we only fix the stats and leave the weight and bias to be leanable. We find that it is slightly better (0.1~0.2) in many cases.
5. For GPU usage, it is normal to have some periodical changes, though 60% may be somewhat low. Are you using NCCL2 as the communication backend?
6. The idea is not to avoid replicating models to all gpus and gathering the outputs for each iteration, which will significantly slow down training.

**Comment by xmyqsh** [comment]
1. When all operations experssed on tensor, computing on GPU, nms may become the bottleneck. Also, the nms_kernel.cu is implemented from the first version of faster-rcnn, but it appears no significant advantage than the cpu version at the time. Anyway, further profiling will give a good answer.
2. The [nms implementation](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/csrc/cuda/nms.cu) in [maskrcnn-benchmark](https://github.com/facebookresearch/maskrcnn-benchmark) is excectly adopted from [faster-rcnn](https://github.com/ShaoqingRen/faster_rcnn/blob/master/functions/nms/nms_gpu_mex.cu), and I have had a conversation with the author of it(Right, he announced he is the author of it). He converts the greedy problem into a binary mask mapping problem, and use gpu solving the mask mapping problem like solving a matrix computation problem, which is gpu good at. I don't think it is possible to express a nms kernel into graph in pytorch.
3. ok
4. I think so.
5. yes, NCCL 2.2
6. Not so clear. For data parallel on GPUs compared with model parallel, it is indeed replicating models to all gpus, gathering the outputs for each iteration, even thought haven't adopted your  Iteration pipeline. I think one process per gpu makes big difference. `torch.distributed.launch` by NCCL allows high performance by overlapping communication with computation during backward. Not so familiar at this part yet.
